sbin/mount_msdos/mount_msdos.8
.\" $NetBSD: mount_msdos.8,v 1.36 2012/11/16 15:00:18 tsutsui Exp $
.\"
.\" Copyright (c) 1993, 1994 Christopher G. Demetriou
.\" All rights reserved.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\" 3. All advertising materials mentioning features or use of this software
.\"    must display the following acknowledgement:
.\"          This product includes software developed for the
.\"          NetBSD Project.  See http://www.NetBSD.org/ for
.\"          information about NetBSD.
.\" 4. The name of the author may not be used to endorse or promote products
.\"    derived from this software without specific prior written permission.
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE AUTHOR ``AS IS'' AND ANY EXPRESS OR
.\" IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES
.\" OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED.
.\" IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY DIRECT, INDIRECT,
.\" INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT
.\" NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
.\" DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
.\" THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
.\" (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF
.\" THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
.\"
.\" <<Id: LICENSE,v 1.2 2000/06/14 15:57:33 cgd Exp>>
.\"
.Dd November 16, 2012
.Dt MOUNT_MSDOS 8
.Os
.Sh NAME
.Nm mount_msdos
.Nd mount an MS-DOS file system
.Sh SYNOPSIS
.Nm
.Op Fl 9Gls
.Op Fl C Ar codepage
.Op Fl g Ar gid
.Op Fl I Ar iocharset
.Op Fl M Ar mask
.Op Fl m Ar mask
.Op Fl o Ar options
.Op Fl t Ar gmtoff
.Op Fl u Ar uid
.Pa special
.Pa node
.Sh DESCRIPTION
The
.Nm
command attaches the MS-DOS filesystem residing on
the device
.Pa special
to the global filesystem namespace at the location
indicated by
.Pa node .
Both
.Ar special
and
.Ar node
are converted to absolute paths before use.
This command is normally executed by
.Xr mount 8
at boot time, but can be used by any user to mount an
MS-DOS file system on any directory that they own (provided,
of course, that they have appropriate access to the device that
contains the file system).
.Pp
Support for FAT16 and VFAT32 as well as long file names is available.
.Pp
The options are as follows:
.Bl -tag -width XoXoptionsXX
.It Fl 9
Ignore the special Win'95 directory entries even
if deleting or renaming a file.
This forces
.Fl s .
.It Fl C Ar codepage
Specify the MS-DOS code page (aka IBM/OEM code page) name used for
file name conversions for DOS names.
.It Fl G
This option causes the filesystem to be interpreted as an Atari-Gemdos
filesystem.
The differences to the MS-DOS filesystem are minimal and
limited to the boot block.
This option also allows mounting X680x0's Human68k floppies.
This option enforces
.Fl s .
.It Fl g Ar gid
Set the group of the files in the file system to
.Ar gid .
The default group is the group of the directory
on which the file system is being mounted.
.It Fl I Ar iocharset
Specify the I/O charset name used for file name conversions.
.It Fl l
Force listing and generation of
Win'95 long filenames
and separate creation/modification/access dates.
.Pp
If neither
.Fl s
nor
.Fl l
are given,
.Nm
searches the root directory of the filesystem to
be mounted for any existing Win'95 long filenames.
If the filesystem is not empty and no such entries are found,
.Fl s
is the default.
Otherwise
.Fl l
is assumed.
.It Fl M Ar mask
Specify the maximum file permissions for directories
in the file system.
The value of
.Fl m
is used if it is supplied and
.Fl M
is omitted.
.It Fl m Ar mask
Specify the maximum file permissions for files
in the file system.
(For example, a mask of
.Li 755
specifies that, by default, the owner should have
read, write, and execute permissions for files, but
others should only have read and execute permissions.
See
.Xr chmod 1
for more information about octal file modes.)
Only the nine low-order bits of
.Ar mask
are used.
The value of
.Fl M
is used if it is supplied and
.Fl m
is omitted.
The default mask is taken from the
directory on which the file system is being mounted.
.It Fl o Ar options
Use the specified mount
.Ar options ,
as described in
.Xr mount 8 .
.It Fl s
Force behaviour to
ignore and not generate Win'95 long filenames.
See also
.Fl l .
.It Fl t Ar gmtoff
Set the time zone offset (in seconds) from UTC to
.Ar gmtoff ,
with positive values indicating east of the Prime Meridian.
If not set, the user's current time zone will be used.
.It Fl u Ar uid
Set the owner of the files in the file system to
.Ar uid .
The default owner is the owner of the directory
on which the file system is being mounted.
.El
.Sh EXAMPLES
To remove the 'execute' permission bit for all files, but still keep
directories searchable, use:
.Bl -item -offset indent
.It
mount_msdos -m 0644 -M 0755 /dev/wd0e /msdos
.El
.Pp
To mount a Japanese MS-DOS file system:
.Bl -item -offset indent
.It
mount_msdos -C cp932 /dev/sd0e /msdos
.El
.Sh SEE ALSO
.Xr mount 2 ,
.Xr unmount 2 ,
.Xr fstab 5 ,
.Xr mount 8
.Pp
List of Localized MS Operating Systems:
.Pa http://msdn.microsoft.com/ja-jp/goglobal/cc563921.aspx
.Sh HISTORY
The
.Nm
utility first appeared in
.Nx 0.9 .
Its predecessor, the
.Ic mount_pcfs
utility appeared in
.Nx 0.8 ,
and was abandoned in favor
of the more aptly-named
.Nm .
.Sh BUGS
Compressed partitions are not supported.
.Pp
The use of the
.Fl 9
flag could result in damaged filesystems,
albeit the damage is in part taken care of by
procedures similar to the ones used in Win'95.


sbin/mount_msdos/mount_msdos.c
/* $NetBSD: mount_msdos.c,v 1.47 2009/10/07 20:34:02 pooka Exp $ */

/*
 * Copyright (c) 1994 Christopher G. Demetriou
 * All rights reserved.
 * 
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 * 3. All advertising materials mentioning features or use of this software
 *    must display the following acknowledgement:
 *          This product includes software developed for the
 *          NetBSD Project.  See http://www.NetBSD.org/ for
 *          information about NetBSD.
 * 4. The name of the author may not be used to endorse or promote products
 *    derived from this software without specific prior written permission.
 * 
 * THIS SOFTWARE IS PROVIDED BY THE AUTHOR ``AS IS'' AND ANY EXPRESS OR
 * IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES
 * OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED.
 * IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY DIRECT, INDIRECT,
 * INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT
 * NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
 * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
 * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
 * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF
 * THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 * 
 * <<Id: LICENSE,v 1.2 2000/06/14 15:57:33 cgd Exp>>
 */

#include <sys/cdefs.h>
#ifndef lint
__RCSID("$NetBSD: mount_msdos.c,v 1.47 2009/10/07 20:34:02 pooka Exp $");
#endif /* not lint */

#include <sys/param.h>
#include <sys/mount.h>
#include <sys/stat.h>
#include <msdosfs/msdosfsmount.h>
#include <err.h>
#include <grp.h>
#include <locale.h>
#include <iconv.h>
#include <langinfo.h>
#include <pwd.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <time.h>
#include <unistd.h>
#include <ctype.h>
#include <util.h>

#include <mntopts.h>

#include "mountprog.h"
#include "mount_msdos.h"

static const struct mntopt mopts[] = {
	MOPT_STDOPTS,
	MOPT_ASYNC,
	MOPT_SYNC,
	MOPT_UPDATE,
	MOPT_GETARGS,
	MOPT_NULL,
};

static void	usage(void) __dead;

#ifndef MOUNT_NOMAIN
int
main(int argc, char **argv)
{

	setprogname(argv[0]);
	return mount_msdos(argc, argv);
}
#endif

void
mount_msdos_parseargs(int argc, char **argv,
	struct msdosfs_args *args, int *mntflags,
	char *canon_dev, char *canon_dir)
{
	struct stat sb;
	int c, set_gid, set_uid, set_mask, set_dirmask, set_gmtoff;
	char *dev, *dir;
	time_t now;
	struct tm *tm;
	mntoptparse_t mp;
	size_t len;

	*mntflags = set_gid = set_uid = set_mask = set_dirmask = set_gmtoff = 0;
	(void)memset(args, '\0', sizeof(*args));

	while ((c = getopt(argc, argv, "Gsl9u:g:m:M:o:t:C:I:")) != -1) {
		switch (c) {
		case 'G':
			args->flags |= MSDOSFSMNT_GEMDOSFS;
			break;
		case 's':
			args->flags |= MSDOSFSMNT_SHORTNAME;
			break;
		case 'l':
			args->flags |= MSDOSFSMNT_LONGNAME;
			break;
		case '9':
			args->flags |= MSDOSFSMNT_NOWIN95;
			break;
		case 'u':
			args->uid = a_uid(optarg);
			set_uid = 1;
			break;
		case 'g':
			args->gid = a_gid(optarg);
			set_gid = 1;
			break;
		case 'm':
			args->mask = a_mask(optarg);
			set_mask = 1;
			break;
		case 'M':
			args->dirmask = a_mask(optarg);
			set_dirmask = 1;
			break;
		case 'o':
			mp = getmntopts(optarg, mopts, mntflags, 0);
			if (mp == NULL)
				err(1, "getmntopts");
			freemntopts(mp);
			break;
		case 't':
			args->gmtoff = atoi(optarg);
			set_gmtoff = 1;
			break;
		case 'C':
			len = strlen(optarg);
			if (len >= sizeof(args->codepage)) {
				fprintf(stderr, "error: codepage(%s) is "
				    "too long.\n", optarg);
				exit(1);
			}
			strlcpy(args->codepage, optarg, sizeof(args->codepage));
			break;
		case 'I':
			len = strlen(optarg);
			if (len >= sizeof(args->iocharset)) {
				fprintf(stderr, "error: iocharset(%s) is "
				    "too long.\n", optarg);
				exit(1);
			}
			strlcpy(args->iocharset, optarg,
			    sizeof(args->iocharset));
			break;
		case '?':
		default:
			usage();
			break;
		}
	}

	if (optind + 2 != argc)
		usage();

	if (set_mask && !set_dirmask) {
		args->dirmask = args->mask;
		set_dirmask = 1;
	} else if (set_dirmask && !set_mask) {
		args->mask = args->dirmask;
		set_mask = 1;
	}

	dev = argv[optind];
	dir = argv[optind + 1];

	pathadj(dev, canon_dev);
	pathadj(dir, canon_dir);

	args->fspec = canon_dev;
	if (!set_gid || !set_uid || !set_mask) {
		if (stat(dir, &sb) == -1)
			err(1, "stat %s", dir);

		if (!set_uid)
			args->uid = sb.st_uid;
		if (!set_gid)
			args->gid = sb.st_gid;
		if (!set_mask) {
			args->mask = args->dirmask =
				sb.st_mode & (S_IRWXU | S_IRWXG | S_IRWXO);
		}
	}

	if (!set_gmtoff) {
		/* use user's time zone as default */
		time(&now);
		tm = localtime(&now);
		args->gmtoff = tm->tm_gmtoff;

	}

	if (args->codepage[0] != '\0' && args->iocharset[0] == '\0') {
		const char *codeset = nl_langinfo(CODESET);
		if (strlen(codeset) >= sizeof(args->iocharset)) {
			fprintf(stderr, "warning: current codeset(%s) "
			    "is too long.\n", codeset);
			goto disabled;
		}
		strlcpy(args->iocharset, codeset, sizeof(args->iocharset));
	}
	if (args->codepage[0] != '\0' && args->iocharset[0] != '\0') {
		iconv_t ic = iconv_open(args->iocharset, args->codepage);
		if (ic == (iconv_t)-1) {
			fprintf(stderr, "warning: unusable codepage(%s) or "
			    "iocharset(%s).\n",
			    args->codepage, args->iocharset);
			goto disabled;
		}
		iconv_close(ic);
	} else if (args->codepage[0] != '\0' || args->iocharset[0] != '\0') {
disabled:
		/* disable filename convertion */
		fprintf(stderr, "warning: disable filename conversion.\n");
		args->codepage[0] = args->iocharset[0] = '\0';
	}

	args->flags |= MSDOSFSMNT_VERSIONED;
	args->version = MSDOSFSMNT_VERSION;
}

int
mount_msdos(int argc, char **argv)
{
	struct msdosfs_args args;
	char canon_dev[MAXPATHLEN], canon_dir[MAXPATHLEN];
	int mntflags;

	setlocale(LC_CTYPE, "");

	mount_msdos_parseargs(argc, argv, &args, &mntflags,
	    canon_dev, canon_dir);

	if (mount(MOUNT_MSDOS, canon_dir, mntflags, &args, sizeof args) == -1)
		err(1, "%s on %s", canon_dev, canon_dir);

	if (mntflags & MNT_GETARGS) {
		char buf[1024];
		(void)snprintb(buf, sizeof(buf), MSDOSFSMNT_BITS, args.flags);
		printf("uid=%d, gid=%d, mask=0%o, dirmask=0%o, gmtoff=%d, "
		    "codepage=%s, codeset=%s, flags=%s\n",
		    args.uid, args.gid, args.mask, args.dirmask, args.gmtoff,
		    args.codepage, args.iocharset, buf);
	}

	exit (0);
}

static void
usage(void)
{

	fprintf(stderr, "usage: %s [-9Gls] [-C codepage] [-g gid] [-M mask]\n"
	    "\t[-m mask] [-o options] [-t gmtoff] [-u uid] special mountpath\n",
	    getprogname());
	exit(1);
}
